### ISSUE TITLE
How to register other flutter plugins inside FlutterEngine on iOS? 

### ISSUE THREAD

**Issue by rafaelsetragni** [flutter/flutter#99883]
Hello!

Im trying to create a background Flutter Engine to execute dart code while the app is terminated. This is necessary to receive user's notification action without bring the app to foreground, for my plugin called [Awesome Notifications](https://pub.dev/packages/awesome_notifications).

But im struggling to discover how to register the plugins on iOS, to allow then to be used inside the new Flutter Engine instance.
How should I register all plugins to be used inside my background Flutter Engine?

## Steps to Reproduce

This is my simplified swift code (the current real implementation can be found [here](https://github.com/rafaelsetragni/awesome_notifications/blob/add0ff600482df2742f1f957efcc14372abb4470/ios/Classes/lib/DartBackgroundExecutor.swift#L100)):
```Swift
DispatchQueue.main.async {
    
    self.backgroundEngine = FlutterEngine(
        name: "AwesomeNotificationsBgEngine",
        project: nil,
        allowHeadlessExecution: true
    )
    
    if self.backgroundEngine == nil {
        Log.d(self.TAG, "Flutter background engine is not available.")
        self.closeBackgroundIsolate()
    }
    else {
        
        self.backgroundEngine!.run(
            withEntrypoint: dartCallbackInfo.callbackName,
            libraryURI: dartCallbackInfo.callbackLibraryPath)                
        
        FlutterPluginRegistrar
            .register( self.backgroundEngine.)
        
        self.initializeReverseMethodChannel(
            backgroundEngine: self.backgroundEngine!)
    }
}
```

...and on dart code:

```Dart
static Future<void> onActionReceivedMethod(ReceivedAction receivedAction) async {
    SharedPreferences prefs = await SharedPreferences.getInstance();
    prefs.setString('stringValue', "abc");
}
```

...it works well on Android, but on iOS im getting this error:
`MissingPluginException(No implementation found for method getAll on channel plugins.flutter.io/shared_preferences)`

But that's not shared_preferences fault, it could be any other plugin in the error, mine included. It's because I'm not registering the plugins on Flutter Engine startup.

## Logs

<details>

<summary>Logs</summary>

```
2022-03-09 21:11:13.101240-0300 Runner[57894:1577666] Metal API Validation Enabled
2022-03-09 21:11:13.293137-0300 Runner[57894:1578362] flutter: Observatory listening on http://127.0.0.1:50265/X7X9IsfkknQ=/
2022-03-09 21:11:13.305895-0300 Runner[57894:1577666] LifeCycleManager subscribe(listener:):77 LiceCycleManager listener successfully attached to iOS
2022-03-09 21:11:13.309764-0300 Runner[57894:1577666] LifeCycleManager currentLifeCycle:138 App is now AppKilled
2022-03-09 21:11:13.325634-0300 Runner[57894:1577666] AwesomeNotificationsPlugin AttachAwesomeNotificationsPlugin(usingRegistrar:throughFlutterChannel:):67 Awesome Notifications plugin attached to iOS 1856.0
2022-03-09 21:11:13.325706-0300 Runner[57894:1577666] AwesomeNotificationsPlugin AttachAwesomeNotificationsPlugin(usingRegistrar:throughFlutterChannel:):68 Awesome Notifications - App Group : group.AwesomeNotifications.me.carda.awesomeNotificationsExample
Engine creation error: Error Domain=com.apple.CoreHaptics Code=-4809 "(null)"
2022-03-09 21:11:13.394346-0300 Runner[57894:1577666] AwesomeNotifications didFinishLaunch(_:):489 Awesome Notifications attached for iOS
2022-03-09 21:11:13.452871-0300 Runner[57894:1577666] LifeCycleManager currentLifeCycle:138 App is now Foreground
2022-03-09 21:11:13.781845-0300 Runner[57894:1577666] AwesomeNotifications initialize(defaultIconPath:channels:backgroundHandle:debug:):395 Awesome Notifications initialized
2022-03-09 21:11:13.781970-0300 Runner[57894:1577666] AwesomeNotificationsPlugin channelMethodInitialize(call:result:):845 Awesome Notifications service initialized
2022-03-09 21:11:13.783210-0300 Runner[57894:1577666] AwesomeEventsReceiver subscribeOnNotificationEvents(listener:):36 <awesome_notifications.AwesomeNotifications: 0x60000185adf0> subscribed to receive notification events
2022-03-09 21:11:13.783372-0300 Runner[57894:1577666] AwesomeEventsReceiver subscribeOnActionEvents(listener:):75 <awesome_notifications.AwesomeNotifications: 0x60000185adf0> subscribed to receive action events
2022-03-09 21:11:13.783430-0300 Runner[57894:1577666] AwesomeNotifications attachAsMainInstance(usingAwesomeEventListener:):123 Awesome notifications 105553141804528 attached to app instance
Notification category identifier: DEFAULT,SNOOZE
2022-03-09 21:17:33.718065-0300 Runner[57894:1577666] BackgroundService enqueue(SilentBackgroundAction:withCompletionHandler:):33 A new Dart background service has started
2022-03-09 21:17:43.178534-0300 Runner[57894:1585586] flutter: Got an unknown Silent Action callback error: MissingPluginException(No implementation found for method getAll on channel plugins.flutter.io/shared_preferences)
2022-03-09 21:17:43.178895-0300 Runner[57894:1585586] flutter:
2022-03-09 21:17:43.179224-0300 Runner[57894:1585586] flutter: Awesome Notifications FCM: An error occurred in your background messaging handler:
2022-03-09 21:17:43.179377-0300 Runner[57894:1585586] flutter: Instance of 'AwesomeNotificationsException'
2022-03-09 21:17:43.182534-0300 Runner[57894:1577666] DartBackgroundExecutor finishDartBackgroundExecution():182 All silent data fetched.
```

```
[✓] Flutter (Channel stable, 2.10.2, on macOS 12.2.1 21D62 darwin-arm, locale en-BR)
    • Flutter version 2.10.2 at /Users/carda/Desenvolvimento/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 097d3313d8 (3 weeks ago), 2022-02-18 19:33:08 -0600
    • Engine revision a83ed0e5e3
    • Dart version 2.16.1
    • DevTools version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/carda/Library/Android/sdk
    • Platform android-32, build-tools 32.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)

[✓] VS Code (version 1.65.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.36.0

[✓] Connected device (2 available)
    • iPhone 13 (mobile) • 6C5C5F32-D01E-4108-B539-A61B00D5DF32 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-15-2 (simulator)
    • Chrome (web)       • chrome                               • web-javascript • Google Chrome 99.0.4844.51

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

**Comment by rafaelsetragni** [comment]
Maybe I just need to get a valid instance of `flutterPluginRegistrantCallback`, but im don't know how I can get this inside a flutter plugin, mainly when the app is terminated.

**Comment by rafaelsetragni** [comment]
Ive got a not so pretty way to register those plugins, changing the AppDelegate.swift file:

```Swift
@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate {
  override func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
  ) -> Bool {
      GeneratedPluginRegistrant.register(with: self)

      SwiftAwesomeNotificationsPlugin.setPluginRegistrantCallback { registry in
          SwiftAwesomeNotificationsPlugin.register(
            with: registry.registrar(forPlugin: "io.flutter.plugins.awesomenotifications.AwesomeNotificationsPlugin")!)          
          FLTSharedPreferencesPlugin.register(
            with: registry.registrar(forPlugin: "io.flutter.plugins.sharedpreferences.SharedPreferencesPlugin")!)
      }

      return super.application(application, didFinishLaunchingWithOptions: launchOptions)
  }
}
```

But honestly, im not enjoying it. I guess it will be confuse to some developers to use it. And only on iOS I have to do it. At least Ill like to implement that registrant process on Android as the same way it is done on iOS. To maintain a patter of how everything works.

**Comment by darshankawar** [comment]
@rafaelsetragni 
See if [this](https://github.com/flutter/flutter/issues/98473#issuecomment-1060952450) issue comment helps.

**Comment by rafaelsetragni** [comment]
Unfortunately,  `DartPluginRegistrant.ensureInitialized()` did not solved the problem. The plugins are still not getting called. 

And i was thinking about it. Is there a way to register the plugins on Dart code, but instead of a global single call, using another method to initialize then individually? Because for small background tasks, not all plugins are needed. This could save computational resources, I guess.

This way I could even remove the Android self registering process to let it even more lightweight.

**Comment by rafaelsetragni** [comment]
Inspecting the Flutter engine source code, I discovered why the plugins are not getting registered in iOS. The source code for Platform.iOS is missing:

![image](https://user-images.githubusercontent.com/40064496/157685235-8bac169d-2d57-4569-80ae-66ef6856cbe3.png)

**Comment by darshankawar** [comment]
@rafaelsetragni 
Thanks for digging in and providing further updates. Did you try the same on latest master ? since per this comment, it has been fixed and merged there: https://github.com/flutter/flutter/issues/98473#issuecomment-1060952450 `For Flutter 2.11+ (currently master), call DartPluginRegistrant.ensureInitialized(),`

This could be related to https://github.com/dart-lang/sdk/issues/44215 

See https://github.com/flutter/flutter/issues/98475#issuecomment-1049180516

**Comment by rafaelsetragni** [comment]
Just a suggestion:

Instead of generate the class GeneratePluginRegistrant like this:
![image](https://user-images.githubusercontent.com/40064496/158034827-2efaec28-bcc3-492c-9eac-43329e462985.png)

... we could generate like this:
![image](https://user-images.githubusercontent.com/40064496/158041933-e5e9b3ba-83b1-47dd-b40c-03515f97055e.png)

This way GeneratedPluginRegistrant can be called using reflection and the plugins can be register later inside isolates by name, individually, using Dart code.

**Comment by darshankawar** [comment]
Thanks for the updated and digging in further.
I am going ahead and keeping it open and labeling it for further insights from the team.

**Comment by rafaelsetragni** [comment]
Put this feature to register the plugins by a Dart method, passing their name as a reference, for Android as well (and even for other platforms, but mainly for Android/iOS). Let's save resources within isolates for all platforms.

**Comment by rafaelsetragni** [comment]
Hi guys, do you have any updates?

**Comment by stuartmorgan** [comment]
Updating title and labels to reflect what's being requested at this point, which as I understand it is a global name->registration method map on the native side, and Dart API to drive per-plugin native registration from the Dart side.

(The suggested code above is showing class names as the name, but it would actually have to be the package name for this to work across platforms.)

**Comment by rafaelsetragni** [comment]
Yes, and will be way more easy to the flutter developers to use. 100% agreed.

**Comment by rafaelsetragni** [comment]
Yes. You need to import the packages that you're intending to use.

**Comment by rafaelsetragni** [comment]
Guys, i found a new issue. There's some plugins that do not have any native implementation and, because of it, im not being able to import the packages and initialize the plugin inside the isolate.

**Comment by stuartmorgan** [comment]
Plugin implementations that are Dart-only are fully handled by `DartPluginRegistrant.ensureInitialized()`.

`path_provider_ios` is not such a plugin though. That error message means you didn't register the native side with your engine instance.

**Comment by stuartmorgan** [comment]
@LangInteger This issue is a request specific to registration of native plugin code; the code in your comment is for registering Dart plugin code.

**Comment by rafaelsetragni** [comment]
All plugins, guys. The devs need to access all plugins, for different reasons, inside isolates. And there's some projects with tons of plugins. So i guess we should provide a way to call the register process for each individual plugin (dart only or native) by name on dart level, to save computational resources on isolate's startup process. This way the developers can only register the plugins that they intent to use inside isolate.


Shared preferences implemented a solution for this case, but all the register process for all plugins already exist on flutter projects. We only need to provide a way to access it by dart code.

**Comment by stuartmorgan** [comment]
> So i guess we should provide a way to call the register process for each individual plugin (dart only or native) by name on dart level

Per-plugin registration of Dart code is already possible. This issue is tracking registration of native code from Dart (see https://github.com/flutter/flutter/issues/99883#issuecomment-1094080338) since that is what doesn't currently exist.

> Shared preferences implemented a solution for this case

No it didn't; the code above only registers the plugin's Dart code.

**Comment by rafaelsetragni** [comment]
How do you register an individual plugin by name on dart code? Do you have an example or doc of how to do it?

**Comment by LangInteger** [comment]
> @LangInteger This issue is a request specific to registration of native plugin code; the code in your comment is for registering Dart plugin code.

@stuartmorgan Thx for your comment. So that's why `DartPluginRegistrant.ensureInitialized()` can be used to register SharedPreference.

The thing I and @rafaelsetragni are confused about, I think, is how can we distinguish a plugin that is dart-only or not dart-only. As far as I know, the SharedPreference plugin depends on different mechanism in iOS and Android, how could it be dart-only. Is it because the native code is located in  `SharedPreferencesAndroid` and `SharedPreferencesIos`, while not in `SharedPreference` itself? Thx in advance.

**Comment by LangInteger** [comment]
Hi @stuartmorgan ,by saying:

> That will only register the Dart component of shared_preferences; the plugin will not work (on most platforms) unless you have also registered the native code with the engine backing the isolate.

You mean with this [solution](https://github.com/flutter/flutter/issues/98473#issuecomment-1060952450):

- For Flutter 2.11+ (currently master), call DartPluginRegistrant.ensureInitialized(), which will run the Dart plugin registration for all plugins that need it.
- For older versions of Flutter, manually call the registerWith of any plugin implementations you use that have Dart registration. For shared_preferences, see https://github.com/flutter/flutter/issues/98473#issuecomment-1041895729 for details.

We still cannot use shared_preferences in dart isolation on iOS or Android (because we still do not register the native code manually with the above code)? I ask this because I think the above quote is the solution to the problem `[shared_preferences] MissingPluginException(No implementation found for method getAll on channel plugins.flutter.io/shared_preferences) in isolate`, which is the title of the [issue #98473](https://github.com/flutter/flutter/issues/98473). Do I miss something here? Much appreciated it if you can clarify this. Thank you.

**Comment by stuartmorgan** [comment]
@LangInteger If you have issues using `shared_preferences` that haven't already been addressed in other issues, please file a new bug; as I said above that's off-topic here.
